crates/ide-diagnostics/src/handlers/incorrect_generics_len.rs

```diff
@@ -0,0 +1,183 @@
+use crate::{Diagnostic, DiagnosticCode, DiagnosticsContext};
+use hir::IncorrectGenericsLenKind;
+
+// Diagnostic: incorrect-generics-len
+//
+// This diagnostic is triggered if the number of generic arguments does not match their declaration.
+pub(crate) fn incorrect_generics_len(
+    ctx: &DiagnosticsContext<'_>,
+    d: &hir::IncorrectGenericsLen,
+) -> Diagnostic {
+    let owner_description = match d.def {
+        hir::GenericDef::Function(_) => "function",
+        hir::GenericDef::Adt(hir::Adt::Struct(_)) => "struct",
+        hir::GenericDef::Adt(hir::Adt::Enum(_)) => "enum",
+        hir::GenericDef::Adt(hir::Adt::Union(_)) => "union",
+        hir::GenericDef::Trait(_) => "trait",
+        hir::GenericDef::TraitAlias(_) => "trait alias",
+        hir::GenericDef::TypeAlias(_) => "type alias",
+        hir::GenericDef::Impl(_) => "impl",
+        hir::GenericDef::Const(_) => "constant",
+        hir::GenericDef::Static(_) => "static",
+    };
```

Comment:
Let's extract that into a method? Seems generally useful for more diagnostics in the future